### PR TITLE
Explicitly binding process.domain to the callback

### PR DESCRIPTION
**Pull request by kedemd**
https://github.com/hapijs/catbox/issues/127

### CODE REVIEW

**Comment by hueniverse** [comment]
What's the performance impact of this change?

**Comment by kedemd** [comment]
The results:
Modified took 8424ms
Original took 8178ms

The code:

```javascript
var Catbox = require('catbox');
var CatboxMemory = require('catbox-memory');
var Hoek = require('./node_modules/catbox/node_modules/hoek');

var Async = require('async');

 require('domain').create().run(function() {

     var originalPolicy = new Catbox.Policy({
         expiresIn: 1000,
         staleIn: 900,
         generateFunc: function (id, next) {
             return next(null, true);
         },
         staleTimeout: 99
     }, new Catbox.Client(CatboxMemory), "#original");
     var modifiedPolicy = new Catbox.Policy({
         expiresIn: 1000,
         staleIn: 900,
         generateFunc: function (id, next) {
             return next(null, true);
         },
         staleTimeout: 99
     }, new Catbox.Client(CatboxMemory), "#modified");

     modifiedPolicy.get = function (key, callback) {     // key: string or { id: 'id' }
         var self = this;

         // Check if request is already pending

         var id = (key && typeof key === 'object') ? key.id : key;
         var pendingsId = '+' + id;                                  // Prefix to avoid conflicts with JS internals (e.g. __proto__)
         if (this._pendings[pendingsId]) {
             this._pendings[pendingsId].push(process.domain.bind(callback));
             return;
         }

         this._pendings[pendingsId] = [process.domain.bind(callback)];

         // Lookup in cache

         var timer = new Hoek.Timer();
         this._get(id, function (err, cached) {

             // Prepare report

             var report = {
                 msec: timer.elapsed(),
                 error: err
             };

             if (cached) {
                 report.stored = cached.stored;
                 report.ttl = cached.ttl;
                 var staleIn = typeof self.rule.staleIn === 'function' ? self.rule.staleIn(cached.stored, cached.ttl) : self.rule.staleIn;
                 cached.isStale = (staleIn ? (Date.now() - cached.stored) >= staleIn : false);
                 report.isStale = cached.isStale;
             }

             // No generate method

             if (!self.rule.generateFunc) {
                 return self._finalize(id, err, cached ? cached.item : null, cached, report);
             }

             // Error / Not found

             if (err || !cached) {

                 return self._generate(id, key, null, report, callback);
             }

             // Check if found and fresh

             if (!cached.isStale) {
                 return self._finalize(id, null, cached.item, cached, report);
             }

             return self._generate(id, key, cached, report, callback);
         });
     };
     originalPolicy.get = function (key, callback) {     // key: string or { id: 'id' }
         var self = this;

         // Check if request is already pending

         var id = (key && typeof key === 'object') ? key.id : key;
         var pendingsId = '+' + id;                                  // Prefix to avoid conflicts with JS internals (e.g. __proto__)
         if (this._pendings[pendingsId]) {
             this._pendings[pendingsId].push(callback);
             return;
         }

         this._pendings[pendingsId] = [callback];

         // Lookup in cache

         var timer = new Hoek.Timer();
         this._get(id, function (err, cached) {

             // Prepare report

             var report = {
                 msec: timer.elapsed(),
                 error: err
             };

             if (cached) {
                 report.stored = cached.stored;
                 report.ttl = cached.ttl;
                 var staleIn = typeof self.rule.staleIn === 'function' ? self.rule.staleIn(cached.stored, cached.ttl) : self.rule.staleIn;
                 cached.isStale = (staleIn ? (Date.now() - cached.stored) >= staleIn : false);
                 report.isStale = cached.isStale;
             }

             // No generate method

             if (!self.rule.generateFunc) {
                 return self._finalize(id, err, cached ? cached.item : null, cached, report);
             }

             // Error / Not found

             if (err || !cached) {

                 return self._generate(id, key, null, report, callback);
             }

             // Check if found and fresh

             if (!cached.isStale) {
                 return self._finalize(id, null, cached.item, cached, report);
             }

             return self._generate(id, key, cached, report, callback);
         });
     };
     
     Async.series([
         function(next) {
             var startTime = Date.now();

             Async.times(100000, function (n, next) {
                 modifiedPolicy.get("", next);
             }, function (err) {
                 console.log("Modified took " + (Date.now() - startTime).toString() + "ms");
                 next();
             });
         },
         function(next) {
             var startTime = Date.now();

             Async.times(100000, function (n, next) {
                 originalPolicy.get("", next);
             }, function (err) {
                 console.log("Original took " + (Date.now() - startTime).toString() + "ms");
                next();
             });
         }], function(err) {
     })
 });

```

**Comment by hueniverse** [comment]
So 3% slower. Seems reasonable for an actual problem.

File: lib/policy.js [hunk 1]
```diff
@@ -44,11 +44,11 @@ internals.Policy.prototype.get = function (key, callback) {     // key: string o
     var id = (key && typeof key === 'object') ? key.id : key;
     var pendingsId = '+' + id;                                  // Prefix to avoid conflicts with JS internals (e.g. __proto__)
     if (this._pendings[pendingsId]) {
-        this._pendings[pendingsId].push(callback);
+        this._pendings[pendingsId].push(process.domain.bind(callback));
```
Comment (hueniverse) [inline lib/policy.js]: I would prefer adding `callback = process.domain.bind(callback);` once around line 41 with a comment explaining why this is done.

**Comment by hueniverse** [comment]
@kedemd Any reason why we can't move the domain binding into the pending state (e.g. only bind the second and on callbacks but not the first)?

**Comment by kedemd** [comment]
I agree and don't see any reason why not.

**Pull request merged**